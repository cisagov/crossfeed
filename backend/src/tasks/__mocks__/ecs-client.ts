import { CommandOptions } from '../ecs-client';

export const runCommand = jest.fn(async (commandOptions: CommandOptions) => {
  return {
    tasks: [
      {
        taskArn: 'mock_task_arn'
      }
    ]
  };
});

export const getNumTasks = jest.fn(() => 0);

export const getLogs = jest.fn(() => 'logs');

export default jest.fn(() => ({
  runCommand,
  getNumTasks,
  getLogs
}));
