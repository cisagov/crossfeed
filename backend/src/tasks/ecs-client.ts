import { ECS, CloudWatchLogs } from 'aws-sdk';
import { SCAN_SCHEMA } from '../api/scans';
import * as Docker from 'dockerode';

export interface CommandOptions {
  organizationId?: string;
  organizationName?: string;
  scanId: string;
  scanName: string;
  scanTaskId: string;
  chunkNumber?: number;
  numChunks?: number;
}

const toSnakeCase = (input) => input.replace(/ /g, '-');

/**
 * ECS Client. Normally, submits jobs to ECS.
 * When the app is running locally, runs a
 * Docker container locally instead.
 */
class ECSClient {
  ecs?: ECS;
  cloudWatchLogs?: CloudWatchLogs;
  docker?: Docker;
  isLocal: boolean;

  constructor(isLocal?: boolean) {
    this.isLocal =
      isLocal ??
      (process.env.IS_OFFLINE || process.env.IS_LOCAL ? true : false);
    if (this.isLocal) {
      this.docker = new Docker();
    } else {
      this.ecs = new ECS();
      this.cloudWatchLogs = new CloudWatchLogs();
    }
  }

  /**
   * Launches an ECS task with the given command.
   * @param commandOptions Command options
   */
  async runCommand(commandOptions: CommandOptions) {
    const {
      scanId,
      scanName,
      organizationId,
      organizationName,
      numChunks,
      chunkNumber
    } = commandOptions;
    const { cpu, memory, global } = SCAN_SCHEMA[scanName];
    if (this.isLocal) {
      try {
        const containerName = toSnakeCase(
          `crossfeed_worker_${
            global ? 'global' : organizationName
          }_${scanName}_` + Math.floor(Math.random() * 10000000)
        );
        const container = await this.docker!.createContainer({
          // We need to create unique container names to avoid conflicts.
          name: containerName,
          Image: 'crossfeed-worker',
          HostConfig: {
            // In order to use the host name "db" to access the database from the
            // crossfeed-worker image, we must launch the Docker container with
            // the Crossfeed backend network.
            NetworkMode: 'crossfeed_backend'
          },
          Env: [
            `CROSSFEED_COMMAND_OPTIONS=${JSON.stringify(commandOptions)}`,
            `DB_DIALECT=${process.env.DB_DIALECT}`,
            `DB_HOST=${process.env.DB_HOST}`,
            `IS_LOCAL=true`,
            `DB_PORT=${process.env.DB_PORT}`,
            `DB_NAME=${process.env.DB_NAME}`,
            `DB_USERNAME=${process.env.DB_USERNAME}`,
            `DB_PASSWORD=${process.env.DB_PASSWORD}`,
            `CENSYS_API_ID=${process.env.CENSYS_API_ID}`,
            `CENSYS_API_SECRET=${process.env.CENSYS_API_SECRET}`,
            `WORKER_USER_AGENT=${process.env.WORKER_USER_AGENT}`,
            `WORKER_SIGNATURE_PUBLIC_KEY=${process.env.WORKER_SIGNATURE_PUBLIC_KEY}`,
            `WORKER_SIGNATURE_PRIVATE_KEY=${process.env.WORKER_SIGNATURE_PRIVATE_KEY}`
          ]
        } as any);
        await container.start();
        return {
          tasks: [
            {
              taskArn: containerName
            }
          ],
          failures: []
        };
      } catch (e) {
        console.error(e);
        return {
          tasks: [],
          failures: [{}]
        };
      }
    }
    const tags = [
      {
        key: 'scanId',
        value: scanId
      },
      {
        key: 'scanName',
        value: scanName
      }
    ];
    if (organizationName && organizationId) {
      tags.push({
        key: 'organizationId',
        value: organizationId
      });
      tags.push({
        key: 'organizationName',
        value: organizationName
      });
    }
    if (numChunks !== undefined && chunkNumber !== undefined) {
      tags.push({
        key: 'numChunks',
        value: String(numChunks)
      });
      tags.push({
        key: 'chunkNumber',
        value: String(chunkNumber)
      });
    }
    return this.ecs!.runTask({
      cluster: process.env.FARGATE_CLUSTER_NAME!,
      taskDefinition: process.env.FARGATE_TASK_DEFINITION_NAME!,
      networkConfiguration: {
        awsvpcConfiguration: {
          assignPublicIp: 'ENABLED',
          securityGroups: [process.env.FARGATE_SG_ID!],
          subnets: [process.env.FARGATE_SUBNET_ID!]
        }
      },
      platformVersion: '1.4.0',
      launchType: 'FARGATE',
      // TODO: enable tags when we are able to opt in to the new ARN format for the lambda IAM role.
      // See https://aws.amazon.com/blogs/compute/migrating-your-amazon-ecs-deployment-to-the-new-arn-and-resource-id-format-2/
      // tags,
      overrides: {
        cpu,
        memory,
        containerOverrides: [
          {
            name: 'main', // from task definition
            environment: [
              {
                name: 'CROSSFEED_COMMAND_OPTIONS',
                value: JSON.stringify(commandOptions)
              },
              {
                name: 'WORKER_USER_AGENT',
                value: process.env.WORKER_USER_AGENT
              },
              {
                // Allow node to use more memory, if needed
                name: 'NODE_OPTIONS',
                value: memory ? `--max_old_space_size=${memory}` : ''
              }
            ]
          }
        ]
      }
    } as ECS.RunTaskRequest).promise();
  }

  /**
   * Gets logs for a specific task.
   */
  async getLogs(fargateTaskArn: string) {
    if (this.isLocal) {
      const logStream = await this.docker?.getContainer(fargateTaskArn).logs({
        stdout: true,
        stderr: true,
        timestamps: true
      });
      // Remove 8 special characters at beginning of Docker logs -- see
      // https://github.com/moby/moby/issues/7375.
      return logStream
        ?.toString()
        .split('\n')
        .map((e) => e.substring(8))
        .join('\n');
    } else {
      const response = await this.cloudWatchLogs!.getLogEvents({
        logGroupName: process.env.FARGATE_LOG_GROUP_NAME!,
        // Pick the ID from "arn:aws:ecs:us-east-1:957221700844:task/f59d71c6-3d23-4ee9-ad68-c7b810bf458b"
        logStreamName: `worker/main/${fargateTaskArn.split('/')[1]}`,
        startFromHead: true
      }).promise();
      const res = response.$response.data;
      if (!res || !res.events?.length) {
        return '';
      }
      return res.events
        .map((e) => `${new Date(e.timestamp!).toISOString()} ${e.message}`)
        .join('\n');
    }
  }

  /**
   * Retrieves the number of running tasks associated
   * with the Fargate worker.
   */
  async getNumTasks() {
    if (this.isLocal) {
      const containers = await this.docker?.listContainers({
        filters: {
          ancestor: ['crossfeed-worker']
        }
      });
      return containers?.length || 0;
    }
    const tasks = await this.ecs
      ?.listTasks({
        cluster: process.env.FARGATE_CLUSTER_NAME,
        launchType: 'FARGATE'
      })
      .promise();
    return tasks?.taskArns?.length || 0;
  }
}

export default ECSClient;
