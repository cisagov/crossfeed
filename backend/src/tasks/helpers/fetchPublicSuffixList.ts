import * as fs from 'fs';

// Checks for PSL at ${path} and fetches a new one if missing or out of date
export default async (path = 'public_suffix_list.dat'): Promise<void> => {
  if (!fs.existsSync(path)) {
    fs.writeFileSync(path, await fetchPSL());
  } else {
    const pslAge = (Date.now() - fs.statSync(path).mtimeMs) / (1000 * 60 * 60);
    if (pslAge > 24) {
      fs.writeFileSync(path, await fetchPSL());
    }
  }
};

async function fetchPSL() {
  return fetch('https://publicsuffix.org/list/public_suffix_list.dat', {
    method: 'GET'
  }).then((res) => res.text());
}
