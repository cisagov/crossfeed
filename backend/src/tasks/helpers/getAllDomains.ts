import { Domain, connectToDatabase } from '../../models';

/** Helper function to fetch all domains */
export default async (): Promise<Domain[]> => {
  await connectToDatabase();

  return Domain.find({
    select: ['id', 'name', 'ip', 'organization']
  });
};
