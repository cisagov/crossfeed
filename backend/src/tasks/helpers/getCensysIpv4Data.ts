// TODO: call censys banner API. https://data-01.censys.io/snapshots/ipv4-banners/20200714/ipv4_banners-20200714-000000001268.json.gz
export default async () => {
  return [];
};
