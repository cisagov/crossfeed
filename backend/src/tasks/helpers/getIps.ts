import { Domain, connectToDatabase } from '../../models';

/** Helper function to fetch all domains that have IPs set. */
export default async (organizationId?: String): Promise<Domain[]> => {
  await connectToDatabase();

  let domains = Domain.createQueryBuilder('domain')
    .leftJoinAndSelect('domain.organization', 'organization')
    .andWhere('ip IS NOT NULL');

  if (organizationId) {
    domains = domains.andWhere('domain.organization=:org', {
      org: organizationId
    });
  }

  return domains.getMany();
};
