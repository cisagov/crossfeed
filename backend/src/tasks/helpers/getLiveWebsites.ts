import { plainToClass } from 'class-transformer';
import { Domain, connectToDatabase, Service } from '../../models';

export class LiveDomain extends Domain {
  url: string;
  service: Service;
}

/** Helper function to fetch all live websites (port 80 or 443) */
export const getLiveWebsites = async (
  organizationId?: string,
  organizationIds: string[] = []
): Promise<LiveDomain[]> => {
  await connectToDatabase();

  if (organizationId) {
    organizationIds = [organizationId];
  }

  if (organizationIds.length === 0) {
    return [];
  }

  const qs = Domain.createQueryBuilder('domain')
    .leftJoinAndSelect('domain.services', 'services')
    .leftJoinAndSelect('domain.organization', 'organization')
    .andWhere('domain.organization IN (:...orgs)', { orgs: organizationIds })
    .groupBy('domain.id, domain.ip, domain.name, organization.id, services.id');

  qs.andHaving(
    "COUNT(CASE WHEN services.port = '443' OR services.port = '80' THEN 1 END) >= 1"
  );

  const websites = await qs.getMany();

  return websites.map((domain) => {
    const live = domain as LiveDomain;
    const ports = domain.services.map((service) => service.port);
    let service: Service;
    if (ports.includes(443))
      service = domain.services.find((service) => service.port === 443)!;
    else service = domain.services.find((service) => service.port === 80)!;
    const url =
      service.port === 443 ? `https://${domain.name}` : `http://${domain.name}`;
    live.url = url;
    live.service = service;
    return live;
  });
};
