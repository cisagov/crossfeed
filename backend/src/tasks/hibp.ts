import { Domain, Service, Vulnerability, connectToDatabase } from '../models';
import { CommandOptions } from './ecs-client';
import got from 'got';
import { plainToClass } from 'class-transformer';
import saveVulnerabilitiesToDb from './helpers/saveVulnerabilitiesToDb';

/**
 * The hibp scan looks up emails from a particular domain
 * that have shown up in breaches using the Have I
 * Been Pwned Enterprise API.
 */
async function getIps(organizationId?: String): Promise<Domain[]> {
  await connectToDatabase();

  let domains = Domain.createQueryBuilder('domain')
    .leftJoinAndSelect('domain.organization', 'organization')
    .andWhere('ip IS NOT NULL');

  if (organizationId) {
    domains = domains.andWhere('domain.organization=:org', {
      org: organizationId
    });
  }
  domains = domains.andWhere('domain.name LIKE :gov', {
    gov: '%.gov'
  });
  domains = domains.andWhere('domain.ipOnly=:bool', {
    bool: false
  });
  return domains.getMany();
}

async function lookupEmails(breachesDict: any, domain: Domain) {
  try {
    const results: any[] = await got(
      'https://haveibeenpwned.com/api/v2/enterprisesubscriber/domainsearch/' +
        domain.name,
      {
        headers: {
          Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
        }
      }
    ).json();


  const AddressResults = {};
  const BreachResults = {};
  const finalResults = {};

    const shouldCountBreach = (breach) =>
      breach.DataClasses.indexOf('Passwords') > -1 &&
      breach.IsVerified === true &&
      breach.BreachDate > '2016-01-01';


    for (const email in results) {
      const filtered = (results[email] || []).filter((e) =>
        shouldCountBreach(breachesDict[e])
      );
      if (filtered.length > 0) {
        AddressResults[email + '@' + domain.name] = filtered;
        for (const breach of filtered) {
          if (!(breach in BreachResults)) {
            BreachResults[breach] = breachesDict[breach];
          }

        }
      }
    }
    finalResults['Emails'] = AddressResults;
    finalResults['Breaches'] = BreachResults;
    return finalResults;
  } catch (error) {
    console.error(
      `Error Occured when trying to access the HIPB API using the domain: ${domain.name} `
    );
    return 0;
  }
}

export const handler = async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running hibp on organization', organizationName);
  const domainsWithIPs = await getIps(organizationId);
  const breaches: any[] = await got(
    'https://haveibeenpwned.com/api/v2/breaches',
    {
      headers: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    }
  ).json();
  const breachesDict = {};
  for (const breach of breaches) {
    breachesDict[breach.Name] = breach;
  }

  const services: Service[] = [];
  const vulns: Vulnerability[] = [];
  for (const domain of domainsWithIPs) {
    console.log(domain.name);
    const results = await lookupEmails(breachesDict, domain);

    if (results) {
      console.log(
        `Got ${Object.keys(results['Emails']).length} emails for domain ${
          domain.name
        }`
      );

      if (Object.keys(results['Emails']).length !== 0) {
        vulns.push(
          plainToClass(Vulnerability, {
            domain: domain,
            lastSeen: new Date(Date.now()),
            title: 'Exposed Emails',
            state: 'open',
            source: 'hibp',
            needsPopulation: false,
            structuredData: {
              emails: results['Emails'],
              breaches: results['Breaches']
            },
            description: `Emails associated with ${domain.name} have been exposed in a breach.`
          })
        );
        await saveVulnerabilitiesToDb(vulns, false);
      }

    }
  }
};
