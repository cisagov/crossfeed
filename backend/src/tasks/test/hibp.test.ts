import { getLiveWebsites, LiveDomain } from '../helpers/getLiveWebsites';
import {
  connectToDatabase,
  Domain,
  Organization,
  Scan,
  Service,
  Vulnerability
} from '../../models';
import { CommandOptions } from '../ecs-client';
import { handler as hibp } from '../hibp';
import * as nock from 'nock';

const RealDate = Date;

const axios = require('axios');

const hibpResponse_1 = {
  testEmail_1: ['Breach_1'],
  testEmail_2: ['Breach_4'],
  testEmail_3: ['Breach_3'],
  testEmail_4: ['Breach_4'],
  testEmail_5: ['Breach_2'],
  testEmail_6: ['Breach_2'],
  testEmail_7: ['Breach_5'],
  testEmail_8: ['Breach_5'],
  testEmail_9: ['Breach_6']
};

const hibpResponse_2 = { testEmail_10: ['Breach_7'] };

const hibpResponse_3 = {};

const breachResponse = [
  {
    Name: 'Breach_1',
    Title: 'Breach_1',
    Domain: 'Breach_1.com',
    BreachDate: '2017-02-01',
    AddedDate: '2017-10-26T23:35:45Z',
    ModifiedDate: '2017-12-10T21:44:27Z',
    PwnCount: 8393093,
    Description: 'Mock Breach number 1',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_1.png',
    DataClasses: ['Email addresses', 'IP addresses', 'Names', 'Passwords'],
    IsVerified: true,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  },
  {
    Name: 'Breach_2',
    Title: 'Breach_2',
    Domain: 'Breach_2.com',
    BreachDate: '2020-03-22',
    AddedDate: '2020-11-15T00:59:50Z',
    ModifiedDate: '2020-11-15T01:07:10Z',
    PwnCount: 8661578,
    Description: 'Mock Breach number 2',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_2.png',
    DataClasses: [
      'Email addresses',
      'IP addresses',
      'Names',
      'Passwords',
      'Phone numbers',
      'Physical addresses',
      'Usernames'
    ],
    IsVerified: true,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  },
  {
    Name: 'Breach_3',
    Title: 'Breach_3',
    Domain: 'Breach_3.com',
    BreachDate: '2012-01-01',
    AddedDate: '2016-10-08T07:46:05Z',
    ModifiedDate: '2016-10-08T07:46:05Z',
    PwnCount: 6414191,
    Description: 'Mock Breach number 3',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_3.png',
    DataClasses: ['Email addresses', 'Passwords'],
    IsVerified: false,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  },
  {
    Name: 'Breach_4',
    Title: 'Breach_4',
    Domain: 'Breach_4.com',
    BreachDate: '2016-04-19',
    AddedDate: '2016-07-08T01:55:03Z',
    ModifiedDate: '2016-07-08T01:55:03Z',
    PwnCount: 4009640,
    Description: 'Mock Breach number 4',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_4.png',
    DataClasses: [
      'Device information',
      'Email addresses',
      'IP addresses',
      'Passwords',
      'Usernames'
    ],
    IsVerified: true,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  },
  {
    Name: 'Breach_5',
    Title: 'Breach_5',
    Domain: 'Breach_5.com',
    BreachDate: '2018-02-29',
    AddedDate: '2018-07-08T01:55:03Z',
    ModifiedDate: '2018-07-08T01:55:03Z',
    PwnCount: 389292,
    Description: 'Mock Breach number 5',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_4.png',
    DataClasses: ['Email addresses', 'IP addresses', 'Passwords', 'Usernames'],
    IsVerified: true,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  },
  {
    Name: 'Breach_6',
    Title: 'Breach_6',
    Domain: 'Breach_6.com',
    BreachDate: '2017-01-22',
    AddedDate: '2017-03-08T01:55:03Z',
    ModifiedDate: '2017-03-08T01:55:03Z',
    PwnCount: 7829238,
    Description: 'Mock Breach number 6',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_4.png',
    DataClasses: ['Device information', 'Email addresses', 'IP addresses'],
    IsVerified: true,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  },
  {
    Name: 'Breach_7',
    Title: 'Breach_7',
    Domain: 'Breach_7.com',
    BreachDate: '2017-01-01',
    AddedDate: '2017-10-08T07:46:05Z',
    ModifiedDate: '2017-10-08T07:46:05Z',
    PwnCount: 3929020,
    Description: 'Mock Breach number 7',
    LogoPath:
      'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_3.png',
    DataClasses: ['Email addresses', 'Passwords'],
    IsVerified: false,
    IsFabricated: false,
    IsSensitive: false,
    IsRetired: false,
    IsSpamList: false
  }
];

describe('hibp', () => {
  let organization;
  let scan;
  let domains: Domain[] = [];
  let connection;
  beforeEach(async () => {
    connection = await connectToDatabase();
    global.Date.now = jest.fn(() => new Date('2019-04-22T10:20:30Z').getTime());
    organization = await Organization.create({
      name: 'test-' + Math.random(),
      rootDomains: ['test-' + Math.random()],
      ipBlocks: [],
      isPassive: false
    }).save();
    scan = await Scan.create({
      name: 'hibp',
      arguments: {},
      frequency: 999
    }).save();
    domains = [
      await Domain.create({
        name: 'test-domain_1.gov',
        ip: '',
        organization
      }).save(),
      await Domain.create({
        name: 'test-domain_2.gov',
        ip: '',
        organization
      }).save(),
      await Domain.create({
        name: 'test-domain_3.gov',
        ip: '',
        organization
      }).save()
    ];

    jest.mock('../helpers/getIps', () => domains);
  });

  afterEach(async () => {
    global.Date = RealDate;
    jest.unmock('../helpers/getIps');
    await connection.close();
  });
  afterAll(async () => {
    nock.cleanAll();
  });
  const checkDomains = async (organization) => {
    const domains = await Domain.find({
      where: { organization },
      relations: ['organization', 'services']
    });
    expect(
      domains
        .sort((a, b) => a.name.localeCompare(b.name))
        .map((e) => ({
          ...e,
          services: e.services.map((s) => ({
            ...s,
            id: null,
            updatedAt: null,
            createdAt: null
          })),
          organization: null,
          id: null,
          updatedAt: null,
          createdAt: null,
          syncedAt: null,
          name: null
        }))
    ).toMatchSnapshot();
    expect(domains.filter((e) => !e.organization).length).toEqual(0);
  };
  test('basic test', async () => {
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_1.gov')
      .reply(200, hibpResponse_1);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/breaches')
      .reply(200, breachResponse);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_2.gov')
      .reply(200, hibpResponse_2);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_3.gov')
      .reply(200, hibpResponse_3);
    await hibp({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    await checkDomains(organization);
  });
  test('creates vulnerability', async () => {
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_1.gov')
      .reply(200, hibpResponse_1);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/breaches')
      .reply(200, breachResponse);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_2.gov')
      .reply(200, hibpResponse_2);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_3.gov')
      .reply(200, hibpResponse_3);
    await hibp({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    const domain = await Domain.findOne({ id: domains[0].id });
    const vulns = await Vulnerability.find({
      domain: domain
    });
    expect(vulns).toHaveLength(1);
    expect(vulns[0].title).toEqual('Exposed Emails');
    expect(vulns[0].source).toEqual('hibp');
  });
  test('updates existing vulnerability', async () => {
    const domain = await Domain.findOne({ id: domains[0].id });
    const service = await Service.create({
      domain,
      port: 443
    }).save();
    const vulnerability = await Vulnerability.create({
      domain,
      cve: null,
      lastSeen: new Date(Date.now()),
      cpe: null,
      title: `Exposed Emails`,
      description: '123',
      state: 'closed',
      structuredData: {
        emails: {
          testEmail_1: ['Breach_1'],
          testEmail_2: ['Breach_2'],
          testEmail_3: ['Breach_3'],
          testEmail_4: ['Breach_4']
        },
        breaches: {
          breach_1: {
            Name: 'Breach_1',
            Title: 'Breach_1',
            Domain: 'Breach_1.com',
            BreachDate: '2017-02-01',
            AddedDate: '2017-10-26T23:35:45Z',
            ModifiedDate: '2017-12-10T21:44:27Z',
            PwnCount: 8393093,
            Description: 'Mock Breach number 1',
            LogoPath:
              'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_1.png',
            DataClasses: [
              'Email addresses',
              'IP addresses',
              'Names',
              'Passwords'
            ],
            IsVerified: true,
            IsFabricated: false,
            IsSensitive: false,
            IsRetired: false,
            IsSpamList: false
          },
          Breach_2: {
            Name: 'Breach_2',
            Title: 'Breach_2',
            Domain: 'Breach_2.com',
            BreachDate: '2020-03-22',
            AddedDate: '2020-11-15T00:59:50Z',
            ModifiedDate: '2020-11-15T01:07:10Z',
            PwnCount: 8661578,
            Description: 'Mock Breach number 2',
            LogoPath:
              'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_2.png',
            DataClasses: [
              'Email addresses',
              'IP addresses',
              'Names',
              'Passwords',
              'Phone numbers',
              'Physical addresses',
              'Usernames'
            ],
            IsVerified: true,
            IsFabricated: false,
            IsSensitive: false,
            IsRetired: false,
            IsSpamList: false
          },
          Breach_3: {
            Name: 'Breach_3',
            Title: 'Breach_3',
            Domain: 'Breach_3.com',
            BreachDate: '2012-01-01',
            AddedDate: '2016-10-08T07:46:05Z',
            ModifiedDate: '2016-10-08T07:46:05Z',
            PwnCount: 6414191,
            Description: 'Mock Breach number 3',
            LogoPath:
              'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_3.png',
            DataClasses: ['Email addresses', 'Passwords'],
            IsVerified: false,
            IsFabricated: false,
            IsSensitive: false,
            IsRetired: false,
            IsSpamList: false
          },
          Breach_4: {
            Name: 'Breach_4',
            Title: 'Breach_4',
            Domain: 'Breach_4.com',
            BreachDate: '2016-04-19',
            AddedDate: '2016-07-08T01:55:03Z',
            ModifiedDate: '2016-07-08T01:55:03Z',
            PwnCount: 4009640,
            Description: 'Mock Breach number 4',
            LogoPath:
              'https://haveibeenpwned.com/Content/Images/PwnedLogos/Breach_4.png',
            DataClasses: [
              'Device information',
              'Email addresses',
              'IP addresses',
              'Passwords',
              'Usernames'
            ],
            IsVerified: true,
            IsFabricated: false,
            IsSensitive: false,
            IsRetired: false,
            IsSpamList: false
          }
        }
      },
      substate: 'remediated',
      source: 'hibp'
    }).save();
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_1.gov')
      .reply(200, hibpResponse_1);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/breaches')
      .reply(200, breachResponse);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_2.gov')
      .reply(200, hibpResponse_2);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_3.gov')
      .reply(200, hibpResponse_3);
    await hibp({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    const vulns = await Vulnerability.find({
      where: { domain },
      relations: ['service']
    });
    expect(vulns).toHaveLength(1);

    // These fields should stay the same
    expect(vulns[0].title).toEqual('Exposed Emails');
    expect(vulns[0].id).toEqual(vulnerability.id);
    expect(vulns[0].cpe).toEqual(vulnerability.cpe);
    expect(vulns[0].state).toEqual(vulnerability.state);
    expect(vulns[0].source).toEqual(vulnerability.source);

    // These fields should be updated
    expect(
      vulns[0].structuredData['emails']['testEmail_2@test-domain_1.gov']
    ).toEqual(['Breach_4']);
    expect(vulns[0].structuredData['breaches']['Breach_1']['PwnCount']).toEqual(
      8393093
    );
    expect(vulns[0].updatedAt).not.toEqual(vulnerability.updatedAt);
  });
  test('verify breaches without password are included', async () => {
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_1.gov')
      .reply(200, hibpResponse_1);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/breaches')
      .reply(200, breachResponse);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_2.gov')
      .reply(200, hibpResponse_2);
    nock('https://haveibeenpwned.com', {
      reqheaders: {
        Authorization: 'Bearer ' + process.env.HIBP_API_KEY!
      }
    })
      .get('/api/v2/enterprisesubscriber/domainsearch/test-domain_3.gov')
      .reply(200, hibpResponse_3);
    await hibp({
      organizationId: organization.id,
      organizationName: 'organizationName',
      scanId: scan.id,
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
    const domain = await Domain.findOne({ id: domains[0].id });
    const vulns = await Vulnerability.find({
      domain: domain
    });
    expect(vulns).toHaveLength(1);
    expect(vulns[0].title).toEqual('Exposed Emails');
    expect(vulns[0].source).toEqual('hibp');
    expect(vulns[0].structuredData['breaches']['Breach_6']).toBeTruthy();
    expect(
      vulns[0].structuredData['breaches']['Breach_6'].passwordIncluded
    ).toEqual(false);
  });
});
