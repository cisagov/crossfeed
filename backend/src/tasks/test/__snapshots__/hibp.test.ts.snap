// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hibp basic test 1`] = `
Array [
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "",
    "name": null,
    "organization": null,
    "reverseName": "test-domain_1",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "",
    "name": null,
    "organization": null,
    "reverseName": "test-domain_2",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
  Object {
    "asn": null,
    "censysCertificatesResults": Object {},
    "cloudHosted": false,
    "country": null,
    "createdAt": null,
    "fromRootDomain": null,
    "id": null,
    "ip": "",
    "name": null,
    "organization": null,
    "reverseName": "test-domain_3",
    "screenshot": null,
    "services": Array [],
    "ssl": null,
    "syncedAt": null,
    "updatedAt": null,
  },
]
`;
